numpy/core/src/multiarray/mapping.c

```diff
@@ -1557,11 +1570,50 @@ _get_field_view(PyArrayObject *arr, PyObject *ind, PyArrayObject **view)
         }
         Py_INCREF(arr);
         if (PyArray_SetBaseObject(*view, (PyObject *)arr) < 0) {
-            Py_DECREF(*view);
-            *view = NULL;
+            goto view_fail;
+        }
+
+        if (force_view) {
+            return 0;
+        }
+
+        /*
```

Comment:
I think everything from here down should go in a helper function with a name like (but better than) `maybe_copy_view(view)`